### ISSUE TITLE
Memory leak in TCP dns serving

### ISSUE THREAD

**Issue by 42wim** [hashicorp/consul#6903]
#### Overview of the Issue

Found a memoryleak in the TCP DNS serving. We were seeing gradual memory build-up on our consul servers, up to 8GB until we restarted the process. (running v1.6.1)

After some debugging the problem seemed to be in `serveTCP`

```
(pprof) text
Showing nodes accounting for 353.40MB, 94.72% of 373.12MB total
Dropped 107 nodes (cum <= 1.87MB)
Showing top 10 nodes out of 71
      flat  flat%   sum%        cum   cum%
   81.51MB 21.85% 21.85%    81.51MB 21.85%  net.newFD
      73MB 19.57% 41.41%       73MB 19.57%  syscall.anyToSockaddr
   57.25MB 15.34% 56.76%    57.25MB 15.34%  github.com/hashicorp/go-msgpack/codec.(*msgpackDecDriver).decodeBytes
   51.50MB 13.80% 70.56%    51.50MB 13.80%  net.sockaddrToTCP
   26.50MB  7.10% 77.66%       38MB 10.19%  github.com/hashicorp/go-immutable-radix.(*Txn).insert
   19.12MB  5.13% 82.79%   228.14MB 61.14%  github.com/miekg/dns.(*Server).serveTCP
   13.50MB  3.62% 86.41%    13.50MB  3.62%  github.com/hashicorp/go-memdb.(*StringFieldIndex).FromObject
   11.50MB  3.08% 89.49%    11.50MB  3.08%  github.com/hashicorp/go-msgpack/codec.(*msgpackDecDriver).decodeString
      10MB  2.68% 92.17%       10MB  2.68%  github.com/hashicorp/go-immutable-radix.(*Txn).writeNode
    9.50MB  2.55% 94.72%   126.76MB 33.97%  github.com/hashicorp/consul/agent/consul/fsm.restoreKV
```
This issue has been mentioned in https://github.com/miekg/dns/issues/807 and closed in https://github.com/miekg/dns/pull/808

A fix can be found in #6748 

#### Reproduction Steps

Setup a consul cluster, add a service with lots of IP addresses so that tcp needs to be used.
Do a DNS lookup of this service for thousands of times.
See memory usage build up.

### Consul info for both Client and Server

Noticed since v1.6.1
Probably introduced with 28c84845bf082f07724217cd198d5f119009417f

**Comment by i0rek** [comment]
Thanks for reporting @42wim! I will have a look soon.

**Comment by 42wim** [comment]
Fix in #6748 has been merged 🎉
